examples/non-fungible/src/contract.rs

```diff
@@ -0,0 +1,384 @@
+    ) -> ExecutionOutcome<Message> {
+        self.remove_nft(&nft).await;
+        if target_account.chain_id == system_api::current_chain_id() {
+            nft.owner = target_account.owner;
+            self.add_nft(nft).await;
+            ExecutionOutcome::default()
+        } else {
+            let message = Message::Transfer {
+                nft,
+                target_account,
+            };
+
+            ExecutionOutcome::default().with_tracked_message(target_account.chain_id, message)
+        }
+    }
+
+    async fn get_nft(&self, token_id: &TokenId) -> Nft {
+        self.nfts
+            .get(token_id)
+            .await
+            .expect("Failure in retrieving NFT")
+            .expect("NFT should not be None")
+    }
+
+    async fn mint(
+        &mut self,
+        owner: AccountOwner,
+        name: String,
+        extra_data: Vec<u8>,
+    ) -> ExecutionOutcome<Message> {
+        let metadata = NftMetadata {
+            name,
+            minter: owner,
+            extra_data,
+        };
+
+        let token_id = Nft::create_token_id(
+            &system_api::current_chain_id(),
+            &system_api::current_application_id(),
+            &metadata,
+        );
+
+        self.add_nft(Nft {
+            token_id: token_id.clone(),
+            owner,
+            metadata,
+        })
+        .await;
+
+        ExecutionOutcome::default()
+    }
+
+    fn remote_claim(
+        &mut self,
+        source_account: Account,
+        token_id: TokenId,
+        target_account: Account,
+    ) -> ExecutionOutcome<Message> {
+        let message = Message::Claim {
+            source_account,
+            token_id,
+            target_account,
+        };
+        ExecutionOutcome::default().with_authenticated_message(source_account.chain_id, message)
+    }
+
+    async fn add_nft(&mut self, nft: Nft) {
+        let token_id = nft.token_id.clone();
+        let owner = nft.owner;
+
+        self.nfts
+            .insert(&token_id, nft)
+            .expect("Error in insert statement");
+        if let Some(owned_token_ids) = self
+            .owners
+            .get_mut(&owner)
+            .await
+            .expect("Error in get_mut statement")
+        {
+            owned_token_ids.insert(token_id);
+        } else {
+            let mut owned_token_ids = BTreeSet::new();
+            owned_token_ids.insert(token_id);
+            self.owners
+                .insert(&owner, owned_token_ids)
+                .expect("Error in insert statement");
+        }
+    }
+
+    async fn remove_nft(&mut self, nft: &Nft) {
+        self.nfts
+            .remove(&nft.token_id)
+            .expect("Failure removing NFT");
+        let owned_token_ids = self
+            .owners
+            .get_mut(&nft.owner)
+            .await
+            .expect("Error in get_mut statement")
+            .expect("NFT set should be there!");
+
+        owned_token_ids.remove(&nft.token_id);
+    }
```

Comment:
Do we care about trying to make these more "atomic"? If we do, I might need to think of another way to implement this

Reply:
This is always part of a transaction